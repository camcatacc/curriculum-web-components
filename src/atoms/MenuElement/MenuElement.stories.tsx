import React, { useState } from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import MenuElement from "atoms/MenuElement/MenuElement";

const Template: ComponentStory<typeof MenuElement> = (args) => {
	const [selected, setSelected] = useState(false);
	const onClicked = () => !selected && setSelected(true);
	return <MenuElement {...args} selected={selected} onClicked={onClicked} />;
};

export const Standard = Template.bind({});
Standard.args = {
	title: "Atoms/Menu element 1",
	color: "black"
};

export default {
	title: "Atoms/MenuElement",
	component: MenuElement
} as ComponentMeta<typeof MenuElement>;
