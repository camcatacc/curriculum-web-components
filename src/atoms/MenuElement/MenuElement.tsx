// Modules
import React from "react";

// Elements
import { Typography } from "@material-ui/core";

// Styles
import * as Styled from "atoms/MenuElement/MenuElement.styled";

// Definitions
export interface MenuElementProps {
	selected: boolean;
	onClicked?: () => void;
	title: string;
	color?: string;
}

// Element
const MenuElement = ({ selected, onClicked, title, color }: MenuElementProps) => {
	const onClick: React.MouseEventHandler<HTMLAnchorElement> & React.MouseEventHandler<HTMLSpanElement> = (e) => {
		e.preventDefault();
		onClicked && onClicked();
	};
	return (
		<Typography variant="body2">
			<Styled.Link
				color="textSecondary"
				$selected={selected}
				$color={color}
				className={`cursor-pointer ${selected ? "font-semibold" : "font-regular"} uppercase`}
				onClick={onClick}
				href="#"
			>
				{title}
			</Styled.Link>
		</Typography>
	);
};
export default MenuElement;
