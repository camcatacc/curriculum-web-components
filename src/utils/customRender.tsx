import { ThemeProvider } from "@material-ui/styles";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ReactElement } from "react";
import { getMaterialTheme } from "styles/materialUi";

export const customRender = (ui: ReactElement, renderOptions?: RenderOptions & { darkMode?: boolean }): RenderResult => {
	const Wrapper: React.FC = ({ children }) => {
		return <ThemeProvider theme={getMaterialTheme(renderOptions?.darkMode ? "dark" : "light")}>{children}</ThemeProvider>;
	};

	return render(ui, { wrapper: Wrapper, ...renderOptions });
};
