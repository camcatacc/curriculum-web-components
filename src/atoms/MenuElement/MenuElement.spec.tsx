// Modules
import React from "react";
import { fireEvent } from "@testing-library/react";
import { customRender } from "utils/customRender";
import "@testing-library/jest-dom/extend-expect";

// Elements
import MenuElement from "atoms/MenuElement/MenuElement";

// Styling
import { primary } from "styles/customColors";

// Definitions
import type { MenuElementProps } from "atoms/MenuElement/MenuElement";

// Default Props
const defaultProps: MenuElementProps = {
	onClicked: jest.fn(),
	selected: false,
	title: "Home",
	color: "white"
};

// Rendering
const renderComponent = (props = {}) => {
	const combinedProps = { ...defaultProps, ...props };
	return customRender(<MenuElement {...combinedProps} />);
};

// Tests
describe("MenuElement", () => {
	it("Renders as expected", () => {
		const { container } = renderComponent();
		expect(container).toMatchSnapshot();
	});

	it("If selected is not true, the color is white", () => {
		const { getByRole } = renderComponent({ selected: false });
		const link = getByRole("link");
		expect(link).toHaveStyle("color: rgba(0, 0, 0, 0.54)");
	});

	it("If selected is true, the color is primary", () => {
		const { getByRole } = renderComponent({ selected: true });
		const link = getByRole("link");
		expect(link).toHaveStyle(`color: ${primary}`);
	});

	it("If clicked, call the onClick function", () => {
		const { getByRole } = renderComponent();
		const link = getByRole("link");
		fireEvent.click(link);
		expect(defaultProps.onClicked).toHaveBeenCalled();
	});
});
