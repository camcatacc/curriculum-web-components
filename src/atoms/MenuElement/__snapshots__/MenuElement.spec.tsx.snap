// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuElement Renders as expected and has no accessibility violations 1`] = `
<div>
  <p
    class="MuiTypography-root MuiTypography-body2"
  >
    <a
      class="MuiTypography-root MuiLink-root MuiLink-underlineHover sc-bdvvtL clPpEY cursor-pointer font-regular uppercase MuiTypography-colorTextSecondary"
      href="#"
    >
      Home
    </a>
  </p>
</div>
`;
